### PR TITLE
Fix grpc middleware interceptor not PostCall-ing when a streaming RPC with non-streaming server finishes successfully.

### PR DESCRIPTION
**Pull request by alexandrupitis1**
Current Client-side Stream Interceptor implementation is bugged - it assumes that if the RPC is streaming, then the Server must be streaming responses too. This is incorrect, as a Streaming RPC can be either a Bidi stream ( which are currently handled properly ), or a Client-side only stream. In this case, if the RPC finished successfully, then PostCall never fires as Interceptor assumes it should get an io.EOF ( which obviously does not make sense when Server is not sending a response stream ).

This most obviously manifests when using it with Prometheus reporter - client side metrics for such RPCs that depend on PostCall ( grpc_client_handled_total, grpc_client_handling_seconds ) are not exported.

## Changes

- Client StreamInterceptor now handles cases when Server is streaming or non-streaming separately.

## Verification

- I created a POC with a Client-streaming, Server non-streaming RPC and modified the example in this repo to run it. Verified that the example does not export certain prometheus metrics ( grpc_client_handled_total, grpc_client_handling_seconds ) for said RPC.
- Applied the fix, verified again that the metrics are now exported

.

### CODE REVIEW

**Review by johanbrandhorst — CHANGES_REQUESTED** [review]
Nice! Could you add a test for this?

**Comment by alexandrupitis1** [comment]
@johanbrandhorst OK i finally had some time to take another pass at this. I added a test and a relevant API to the testpb proto. I had to update grpc-go though - is it OK or should I make an effort to make sure I re-gen the testpb grpc using the same versions you have in repo?

**Review by johanbrandhorst — CHANGES_REQUESTED** [review]

File: testing/testpb/v1/test.proto [hunk 1]
```diff
@@ -65,4 +65,6 @@ service TestService {
 
   rpc PingStream(stream PingStreamRequest) returns (stream PingStreamResponse) {}
 
+  rpc PingStreamSingleResponse(stream PingStreamRequest) returns (PingStreamResponse) {}
```
[the remark this answers is not in the document]
Comment (alexandrupitis1) [reply testing/testpb/v1/test.proto]: Done

**Comment by alexandrupitis1** [comment]
I ran make lint and cleared out all linting issues, but the github check is failing:

>> lint proto files
>> ensuring copyright headers
cannot set copyright headers: you have unstaged changes.
M	README.md

Not exactly sure what to make of it. Do I need to modify something else?

**Comment by johanbrandhorst** [comment]
I think the README thing is a bug, I'm not sure why it happens but I won't let it block this merge

**Review by johanbrandhorst — APPROVED** [review]

**Comment by johanbrandhorst** [comment]
Thank you for your contribution!

**Pull request merged**